src/java.base/share/classes/java/text/MessageFormat.java

```diff
@@ -567,72 +709,79 @@ public String toPattern() {
+            }
+            // Check SubformatPattern
+            if (fmt instanceof DecimalFormat dFmt) {
+                // Quote eligible mFmt pattern characters: '{' and '}'
+                // Here, and in other subformatPattern instances
+                return ",number," + copyAndQuoteBraces(dFmt.toPattern());
+            } else if (fmt instanceof ChoiceFormat cFmt) {
+                return ",choice," + copyAndQuoteBraces(cFmt.toPattern());
+            }
+        } else if (fmt instanceof DateFormat) {
+            // Check dFmt factory instances
+            for (DateFormat.Style style : DateFormat.Style.values()) {
+                if (fmt.equals(DateFormat.getDateInstance(style.getValue(), locale))) {
+                    return ",date" + ((style.getValue() != DateFormat.DEFAULT)
+                            ? "," + style.name().toLowerCase(Locale.ROOT) : "");
                 }
-                if (index >= DATE_TIME_MODIFIERS.length) {
-                    if (fmt instanceof SimpleDateFormat sdfmt) {
-                        result.append(",date");
-                        subformatPattern = sdfmt.toPattern();
-                    } else {
-                        // UNKNOWN
-                    }
-                } else if (index != MODIFIER_DEFAULT) {
-                    result.append(',').append(DATE_TIME_MODIFIER_KEYWORDS[index]);
+                if (fmt.equals(DateFormat.getTimeInstance(style.getValue(), locale))) {
+                    return ",time" + ((style.getValue() != DateFormat.DEFAULT)
+                            ? "," + style.name().toLowerCase(Locale.ROOT) : "");
                 }
-            } else {
-                //result.append(", unknown");
             }
-            if (subformatPattern != null) {
-                result.append(',');
-
-                // The subformat pattern comes already quoted, but only for those characters that are
-                // special to the subformat. Therefore, we may need to quote additional characters.
-                // The ones we care about at the MessageFormat level are '{' and '}'.
-                copyAndQuoteBraces(subformatPattern, result);
+            // Check SubformatPattern
+            if (fmt instanceof SimpleDateFormat sdFmt) {
+                return ",date," + copyAndQuoteBraces(sdFmt.toPattern());
+            }
+        } else if (fmt instanceof ListFormat) {
+            // Check lFmt factory instances
+            for (ListFormat.Type type : ListFormat.Type.values()) {
+                if (fmt.equals(ListFormat.getInstance(locale, type, ListFormat.Style.FULL))) {
+                    return ",list" + ((type != ListFormat.Type.STANDARD)
+                            ? "," + type.name().toLowerCase(Locale.ROOT) : "");
+                }
             }
-            result.append('}');
         }
-        copyAndFixQuotes(pattern, lastOffset, pattern.length(), result);
-        return result.toString();
+        else if (fmt != null) {
```

Comment:
This `else if` does not seem necessary

Reply:
Originally this was needed to make patterns for ClassicFormat (since we could not instanceof check it due to visibility), but as ClassicFormat/DateTimeFormatter cannot be equality checked, this isn't needed. Removed, thanks for spotting.